src/vos/vos_pool.c

```diff
@@ -1542,51 +1597,92 @@ lock_pool_memory(struct vos_pool *pool)
 		pool->vp_umm.umm_base);
 }
 
-/*
- * If successful, this function consumes ph, and closes it upon any error.
- * So the caller shall not close ph in any case.
- */
 static int
-pool_open(void *ph, struct vos_pool_df *pool_df, unsigned int flags, void *metrics,
-	  daos_handle_t *poh)
+pool_open_prep(uuid_t uuid, unsigned int flags, struct vos_pool **p_pool)
 {
 	struct vos_pool		*pool = NULL;
-	struct umem_attr	*uma;
 	struct d_uuid		 ukey;
 	int			 rc;
+	bool			 show_all = true;
 
 	/* Create a new handle during open */
-	rc = pool_alloc(pool_df->pd_id, &pool); /* returned with refcount=1 */
+	rc = pool_alloc(uuid, &pool); /* returned with refcount=1 */
 	if (rc != 0) {
 		D_ERROR("Error allocating pool handle\n");
-		vos_pmemobj_close(ph);
 		return rc;
 	}
 
+	pool->vp_opening = 1;
+	pool->vp_sysdb = !!(flags & VOS_POF_SYSDB);
+	pool->vp_excl = !!(flags & VOS_POF_EXCL);
+	pool->vp_small = !!(flags & VOS_POF_SMALL);
+	pool->vp_rdb = !!(flags & VOS_POF_RDB);
+
+	/*
+	 * Insert the pool into the uuid hash table before full opened, because subsequent
+	 * pool_open process maybe yield and other pool_lookup for the same UUID will find
+	 * the in-processing pool_open, then wait there and avoid concurrent pool_open.
+	 */
+	uuid_copy(ukey.uuid, uuid);
+	rc = d_uhash_link_insert(vos_pool_hhash_get(pool->vp_sysdb), &ukey, &show_all,
+				 &pool->vp_hlink);
+	if (rc != 0) {
+		D_ERROR("uuid hash table insert for pool " DF_UUID " failed: " DF_RC "\n",
+			DP_UUID(uuid), DP_RC(rc));
+		pool_free(pool);
+	} else {
+		/* Here, two references on the pool: one is for myself, the other for the hash. */
+		*p_pool = pool;
+	}
+
+	return rc;
+}
+
+static int
+pool_open_post(struct umem_pool **p_ph, struct vos_pool_df *pool_df, unsigned int flags,
+	       void *metrics, struct vos_pool *pool, int ret)
+{
+	struct umem_attr	*uma;
+	int			 rc;
+
+	if (ret != 0)
+		D_GOTO(out, rc = ret);
+
+	D_ASSERT(pool_df != NULL);
+
 	uma = &pool->vp_uma;
-	uma->uma_pool = ph;
+	uma->uma_pool = *p_ph;
 	uma->uma_id = umempobj_backend_type2class_id(uma->uma_pool->up_store.store_type);
 
+	/* initialize a umem instance for later btree operations */
+	rc = umem_class_init(uma, &pool->vp_umm);
+	if (rc != 0)
+		goto out;
```

Comment:
The vos_pmemobj_close() will be called twice: on pool_hop_free() and on error cleanup code path (when 'p_ph' not set to NULL)?

Reply:
oh, indeed.